core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java

```diff
@@ -814,9 +814,13 @@ public NativeImageInvokerInfo build() {
                             "-H:BuildOutputJSONFile=" + nativeImageName + "-build-output-stats.json");
                 }
 
-                // only available in GraalVM 23.1.0+. Expected to become the default in GraalVM 24.0.0.
+                // only available in GraalVM 23.1.0+.
                 if (graalVMVersion.compareTo(GraalVM.Version.VERSION_23_1_0) >= 0) {
+                    // Expected to become the default in GraalVM 24.0.0.
                     nativeImageArgs.add("--strict-image-heap");
+                    // This feature will become available as non-experimental in the next release (GraalVM for Java 22,
+                    // GraalVM 24.0.0), and will be promoted to the default behavior as the community adopts it.
+                    addExperimentalVMOption(nativeImageArgs, "-H:ThrowMissingRegistrationErrors=");
```

Comment:
I am afraid this change could break downstream code. I would put a package filter that handles only Quarkus classes and Quarkus-managed libraries in code that goes to users. Then, we can ask the users in the next release to add `--throw-missing-registration-errors=` to their build.

We can use the unfiltered version for internal tests (ideally used with our dev builds). In tests, it is ideally used in combination with `-H:MissingRegistrationReportingMode=Exit` to make sure metadata is never missed.

Reply:
Thanks for the input @vjovanov.

FYI when running without `-H:MissingRegistrationReportingMode=Exit` the Quarkus binary exits with error code 1 without printing anything.

When adding `-H:MissingRegistrationReportingMode=Exit` I get to see the exception:

```
org.graalvm.nativeimage.MissingReflectionRegistrationError: The program tried to reflectively access method sun.security.provider.NativePRNG#<init>(java.security.SecureRandomParameters) without it being registered for runtime reflection. Add sun.security.provider.NativePRNG#<init>(java.security.SecureRandomParameters) to the reflection metadata to solve this problem. See https://www.graalvm.org/latest/reference-manual/native-image/metadata/#reflection for help.
```

My understanding is that Quarkus catches the exception at some point which results in the error message not being shown, yet it still fails.

Reply:
>My understanding is that Quarkus catches the exception at some point which results in the error message not being shown, yet it still fails.
I would say the same. That is why we introduced `-H:MissingRegistrationReportingMode=Exit`. 
To speed up the migration you can run the programs with `-H:MissingRegistrationReportingMode=Warn` and it will only print the missing elements.